### PR TITLE
~WIP: first attempt at import list search

### PR DESCRIPTION
**Pull request by bobbyrauchenberg**

### CODE REVIEW

**Comment by bobbyrauchenberg** [comment]
this is branched off `thumphries:topic/filter`

**Comment by thumphries** [comment]
Yeah, you're on the right track! That's a useful one. Good job figuring out all that lens magic.

Would be great to see things like
- export lists
- finishing `findTypeUses` (it doesn't work most of the time)
- `findValueUses` should probably exist, that's the really big one!

I can merge these PRs into the `topic/filter` branch as you go, if you'd prefer that. Just let me know when things get out of WIP.

File: src/Language/Haskell/HGrep/Query.hs [hunk 1]
```diff
@@ -1,10 +1,13 @@
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
```
Comment (thumphries) [inline src/Language/Haskell/HGrep/Query.hs]: Is this being used? If you're just using it for interactive development, you could put it in your ghci.conf https://github.com/thumphries/dotfiles/blob/b3e7cbb7ca36e4c64c68e0409dc208ee3dd24a57/ghc/.ghc/ghci.conf

File: src/Language/Haskell/HGrep/Query.hs [hunk 2]
```diff
@@ -59,16 +70,27 @@ typeContains name ty =
...
 matchDecls src patterns =
   matchDecls' src $ \decl ->
     fromMaybe False (preview patterns decl)
 
+matchDeclsHs :: ParsedSource -> Getting (First Bool) (GHC.ImportDecl RdrName.RdrName) Bool -> [SearchResult]
+matchDeclsHs src patterns =
+  matchDeclsHs' src $ \m ->
+    fromMaybe False (preview patterns m)
+
+matchDeclsHs' :: ParsedSource -> (GHC.ImportDecl RdrName.RdrName -> Bool) -> [SearchResult]
```
Comment (thumphries) [inline src/Language/Haskell/HGrep/Query.hs]: This code becomes a lot less overwhelming when you substitute the type synonyms above! I should have done that myself to make it more readable. One for `ImportDecl` makes sense.

File: src/Language/Haskell/HGrep/Query.hs [hunk 3]
```diff
@@ -48,6 +53,12 @@ findTypeUses name src =
         . _unloc . _ConDeclH98 . _4 . _PrefixCon . folded
         . _unloc . to (typeContains name)
 
+
+findImports :: [Char] -> ParsedSource -> [SearchResult]
+findImports name src =
+  matchDeclsHs src $
+     _ImportDecl . _2 . _unloc . to (cn name)
```
Comment (thumphries) [inline src/Language/Haskell/HGrep/Query.hs]: Curious if this is matching on the _module_ being imported, or the list of _values_ being imported? (We want two prisms, one for each!)

File: src/Language/Haskell/HGrep/Query.hs (hunk 1 above)
Comment (bobbyrauchenberg) [reply src/Language/Haskell/HGrep/Query.hs]: no, i probably shouldn't have pushed this without cleaning it up a bit more. apologies, it was pretty late and i was tired.

that's a good tip about the ghci.conf tho, i didn't know about that! thanks

Comment (thumphries) [reply src/Language/Haskell/HGrep/Query.hs]: For WIP branches, you don't need to clean it up. Any time you want feedback, feel free to open a PR and ask for it.

**Comment by bobbyrauchenberg** [comment]
@thumphries this can probably be merged back into `thumphries:topic/filter`. I've not had time to do anything so far this week, but hopefully might get a bit of time over the weekend

**Comment by thumphries** [comment]
Thanks! I'll be back on this project soon too, I hope :) much further to go

**Pull request merged**